pkg/controllers/pushsecret/pushsecret_controller_template.go

```diff
@@ -0,0 +1,126 @@
+*/
+
+package pushsecret
+
+import (
+	"context"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/external-secrets/external-secrets/apis/externalsecrets/v1alpha1"
+	esv1beta1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+	"github.com/external-secrets/external-secrets/pkg/controllers/templating"
+	_ "github.com/external-secrets/external-secrets/pkg/provider/register" // Loading registered providers.
+	"github.com/external-secrets/external-secrets/pkg/template"
+	"github.com/external-secrets/external-secrets/pkg/utils"
+)
+
+const (
+	errFetchTplFrom = "error fetching templateFrom data: %w"
+	errExecTpl      = "could not execute template: %w"
+)
+
+// merge template in the following order:
+// * template.Data (highest precedence)
+// * template.templateFrom
+// * secret via es.data or es.dataFrom.
+// Whatever is in the Secret THAT'S the Data.
+func (r *Reconciler) applyTemplate(ctx context.Context, ps *v1alpha1.PushSecret, secret *v1.Secret) error {
+	if err := setMetadata(secret, ps); err != nil {
+		return err
+	}
+
+	// no template: copy data and return
+	if ps.Spec.Template == nil {
+		return nil
+	}
+
+	execute, err := template.EngineForVersion(esv1beta1.TemplateEngineV2)
+	if err != nil {
+		return err
+	}
+
+	p := templating.Parser{
+		Client:       r.Client,
+		TargetSecret: secret,
+		DataMap:      secret.Data,
+		Exec:         execute,
+	}
+
+	// apply templates defined in template.templateFrom
+	err = p.MergeTemplateFrom(ctx, ps.Namespace, ps.Spec.Template)
+	if err != nil {
+		return fmt.Errorf(errFetchTplFrom, err)
+	}
+	// explicitly defined template.Data takes precedence over templateFrom
+	err = p.MergeMap(ps.Spec.Template.Data, esv1beta1.TemplateTargetData)
+	if err != nil {
+		return fmt.Errorf(errExecTpl, err)
+	}
+
+	// get template data for labels
+	err = p.MergeMap(ps.Spec.Template.Metadata.Labels, esv1beta1.TemplateTargetLabels)
+	if err != nil {
+		return fmt.Errorf(errExecTpl, err)
+	}
+	// get template data for annotations
+	err = p.MergeMap(ps.Spec.Template.Metadata.Annotations, esv1beta1.TemplateTargetAnnotations)
+	if err != nil {
+		return fmt.Errorf(errExecTpl, err)
+	}
+
+	return nil
+}
+
+// setMetadata sets Labels and Annotations to the given secret.
+func setMetadata(secret *v1.Secret, es *v1alpha1.PushSecret) error {
+	if secret.Labels == nil {
+		secret.Labels = make(map[string]string)
+	}
+	if secret.Annotations == nil {
+		secret.Annotations = make(map[string]string)
+	}
+	// Clean up Labels and Annotations added by the operator
+	// so that it won't leave outdated ones
+	labelKeys, err := templating.GetManagedLabelKeys(secret, es.Name)
+	if err != nil {
+		return err
+	}
+	for _, key := range labelKeys {
+		delete(secret.ObjectMeta.Labels, key)
+	}
+
+	annotationKeys, err := templating.GetManagedAnnotationKeys(secret, es.Name)
+	if err != nil {
+		return err
+	}
+	for _, key := range annotationKeys {
+		delete(secret.ObjectMeta.Annotations, key)
+	}
+
```

Comment:
In the PushSecret case we use the secret as a source and don't manage the labels/annotations, hence i don't think we need to care about the managed fields. That'll simplify this function :)

Reply:
Ah thanks for that context. 🙇

Reply:
Ah now it fully landed. Since we never write it back we just alter it in memory so we can push the altered data to the provider. But we never change the source secret of course. Now this makes sense. :)